app/com/linkedin/drelephant/spark/SparkMetricsAggregator.scala

```diff
@@ -122,6 +128,111 @@ class SparkMetricsAggregator(private val aggregatorConfigurationData: Aggregator
+
+    val sortedStages = notSkippedStages.sortBy { _.submissionTime }
+    var completionTimeWithoutDelayMap: scala.collection.mutable.Map[Int, Long] = scala.collection.mutable.Map[Int, Long]()
+    var completionTimeWithDelayMap: scala.collection.mutable.Map[Int, Long] = scala.collection.mutable.Map[Int, Long]()
+    var finalStageCompletionTimeWithoutDelay = 0L;
+    var finalStageActualCompletionTime = 0L;
+
+    for (i <- 0 until sortedStages.length) {
+      val stage = sortedStages(i)
+      val stageId = stage.stageId
+      var maxCompletionTimeWithoutDelay: Long = 0L;
+      var maxCompletionWithDelay: Long = 0L;
+
+      for (dependeeStageId <- dependencyMap.get(stageId).get) {
+        maxCompletionTimeWithoutDelay = Math.max(maxCompletionTimeWithoutDelay, completionTimeWithoutDelayMap.getOrElse(dependeeStageId, 0L));
+        maxCompletionWithDelay = Math.max(maxCompletionWithDelay, completionTimeWithDelayMap.getOrElse(dependeeStageId, 0L));
+      }
+      if (maxCompletionTimeWithoutDelay == 0) {
+        maxCompletionTimeWithoutDelay = stage.submissionTime.get.getTime;
+        maxCompletionWithDelay = stage.submissionTime.get.getTime
+      }
+      val currentStageDelay = getStageDelay(stage)
+      val completionTimeWithoutDelay = maxCompletionTimeWithoutDelay + (stage.completionTime.get.getTime - stage.submissionTime.get.getTime) - currentStageDelay + (stage.submissionTime.get.getTime - maxCompletionWithDelay)
+
+      logger.debug("currentStageDelay for " + stageId + ":" + currentStageDelay)
+      logger.debug("Delay till this stage " + stageId + ":" + (stage.completionTime.get.getTime - completionTimeWithoutDelay))
+
+      completionTimeWithoutDelayMap.put(stageId, completionTimeWithoutDelay)
+      completionTimeWithDelayMap.put(stageId, stage.completionTime.get.getTime)
+
+      finalStageCompletionTimeWithoutDelay = Math.max(finalStageCompletionTimeWithoutDelay, completionTimeWithoutDelay);
+      finalStageActualCompletionTime = Math.max(finalStageActualCompletionTime, stage.completionTime.get.getTime)
+    }
+
+    logger.debug("Total Delay:" + (finalStageActualCompletionTime - finalStageCompletionTimeWithoutDelay))
+    finalStageActualCompletionTime - finalStageCompletionTimeWithoutDelay
+  }
+
+  /**
+   * Returns scheduler delay for the job, which is max of scheduler delay distribution
+   */
+  def getStageDelay(stage: StageData): Long = {
+    val maxSchedulerDelayIndex = 4
+    var schedulerDelay = 0D
+    val taskSummary = stage.taskSummary.getOrElse(null)
+    if (taskSummary != null) {
+      val schedulerDelays = taskSummary.schedulerDelay
+      if (schedulerDelays != null)
+        schedulerDelay = schedulerDelays(maxSchedulerDelayIndex)
+    }
+    schedulerDelay.longValue()
+  }
+
+  /**
+   * Get approximate dependency between each stages, using submission time and completion time
+   * A stage is dependent on other stage if it is started within 1 second of any stage's completion.
```

Comment:
Nit : I think ,we can put todo here to further improve the dependency logic (instead of relying on 1 second time duration)